libraries/WiFi/src/WiFiGeneric.cpp

```diff
@@ -1450,35 +1448,82 @@ static void wifi_dns_found_callback(const char *name, const ip_addr_t *ipaddr, v
 }
 
 /**
- * Resolve the given hostname to an IP address. If passed hostname is an IP address, it will be parsed into IPAddress structure.
- * @param aHostname     Name to be resolved or string containing IP address
+ * IPv6 compatible DNS callback
+ * @param name
+ * @param ipaddr
+ * @param callback_arg
+ */
+static void wifi_dns6_found_callback(const char *name, const ip_addr_t *ipaddr, void *callback_arg)
+{
+    struct dns_api_msg *msg = (struct dns_api_msg *)callback_arg;
+
+    if(ipaddr && !msg->result) {
+        msg->ip_addr = *ipaddr;
+        msg->result = 1;
+    } else {
+        msg->result = -1;
+    }
+    xEventGroupSetBits(_arduino_event_group, WIFI_DNS_DONE_BIT);
+}
+
+/**
+ * Resolve the given hostname to an IP address.
+ * @param aHostname     Name to be resolved
  * @param aResult       IPAddress structure to store the returned IP address
  * @return 1 if aIPAddrString was successfully converted to an IP address,
  *          else error code
  */
 int WiFiGenericClass::hostByName(const char* aHostname, IPAddress& aResult)
```

Comment:
Once IPAddress is generic, and can handle both 4 and 6, then you only need one version of this function (this name, but with IPv6 support), which will return whichever address is available (and don't need a separate hostByName6 method).

Reply:
Maybe good to have an optional 3rd parameter (enum class type) to set what kind of address you want.
The default being IPv6 if available, otherwise IPv4. (if IPv6 is enabled)

The reason one may explicitly want an IPv4 address could be for legacy reasons where a value needs to be stored and only 4 bytes are available for this.

Reply:
I'm afraid we have legacy here. Many of the existing APIs expect an IPv4, and code using it will crash badly if it's not. I don't think that adding a parameter will solve the backwards compatibility issue. That's why the hostByName6 exists, so if you accept v6, you should use the latter.

Reply:
The default value of such a parameter may be debatable.
However if a check like this is present in the code, you can simply have the default preferred IP type set to IPv4.
```c++
if (WiFiGenericClass::getStatusBits() & WIFI_WANT_IP6_BIT)
```

Reply:
> why the hostByName6 exists

It looks like hostByName6 actually returns both IPv6 and IPv4 addresses (i.e. IPv6 if available, and IPv4 if not, as per RFC 6724).

So maybe they are just badly named, i.e. I would call the generic (return any IP) hostByName(), and then call one that only did IPv4 as hostByName4().

But I guess for legacy if there is a bunch of code that is using hostByName() and then sticking the result in a 4 byte storage then we probably do need to keep the name, no matter how bad it is.

In that case, the dual stack / any version (whatever is available; it's not really dual stack) probably still shouldn't be called with a '6', which is misleading.

Either call it hostByName2(), to simply show it is the newer version (that supersedes the earlier), or call it something like hostByNameDS() or hostAddressByName().  (Really, the whole issue about calling things 'New' or slight variations, e.g. DS, is what happens when you get a third, or fourth 'ReallyNew'... hence suffix 2, 3,... is a generic way that avoids these).


If we need to keep hostByName() as IPv4 only, for legacy, then I would recommend hostByName2() as the new version, that returns the 'best' address, i.e. IPv6 if available, or IPv4 if not.

The '2' suffix makes it clear it is a new version, that should be used for new development; code that uses it needs to handle that it may sometimes return an IPv6 address; and, if IPv6 is turned off, it is backwards compatible. i.e. new code, even if there is no current IPv4 planned, should still be using it, just in case (knowing it may have IPv6).

The '6' suffix is misleading.

Reply:
Actually, that doesn't make sense.

If you are doing stuff that doesn't support IPv6, like sticking values into 4 byte fields, wouldn't you just compile with IPv6 turned off. Then it would just work exactly the same as before, and there would be no compatibility issue.

Then, when you do find/fix whatever the limitation was, all you need to do is turn IPv6 on, and the rest of your code will work, without having to go around everywhere and change references.

Meanwhile, for the usual case, where the value is opaque, i.e. you just do a host lookup to get an IP address, then use that IP address to connect, without caring what is inside that value, then that code can take advantage of IPv6 without having to make any code change. The just use hostByName() and pass the result to connect(). (Without having to go through their code base and change hostByName() to hostByName2() everywhere). The only difference is they use slightly more storage.

They won't know, or care, if they are in an IPv6, dual stack, or IPv4 environment.

The only case where you would be looking inside addressess is for diagnostics, and there you might care about displaying what you got (like "hey, why is it giving me IPv4 when I should have IPv6 available"), and you might want a way to get back multiple addresses. But you don't want to have to write different code for IPv6 vs IPv4.

Reply:
> we have legacy here. Many of the existing APIs expect an IPv4, and code using it will crash badly if it's not.

Yeah, if your code is going to crash with IPv6, then don't turn it on. Use hostByName() (no suffix/prefix) and compile with only IPv4 enabled.

When you do update the legacy code, so that is supports IPv6, then continue to use hostByName() and compile with IPv6.